File: evaluation/S2/python/Hard/2120-kth-smallest-product-of-two-sorted-arrays.py
# Source: https://leetcode.com/problems/kth-smallest-product-of-two-sorted-arrays/   |   Difficulty: Hard
#
# Problem Description:
# Given two sorted 0-indexed integer arrays nums1 and nums2 as well as an integer k, return the kth (1-based) smallest product of nums1[i] * nums2[j] where 0 <= i < nums1.length and 0 <= j < nums2.length.
#
# Example:
# Input: nums1 = [2,5], nums2 = [3,4], k = 2
# Output: 8
# Explanation: The 2 smallest products are:
# - nums1[0] * nums2[0] = 2 * 3 = 6
# - nums1[0] * nums2[1] = 2 * 4 = 8
# The 2nd smallest product is 8.
#
# Constraints:
# 1 <= nums1.length, nums2.length <= 5 * 104
# 	-105 <= nums1[i], nums2[j] <= 105
# 	1 <= k <= nums1.length * nums2.length
# 	nums1 and nums2 are sorted.
#
# Helpful references (internal KB):
# - N/A
#
# Notes:
# - Keep the given function signature unchanged.
# - Use only the Python standard library; no third-party packages.
# - Do not change input/output formats or return value semantics as defined by the template/problem.
# - Conformance: use features compatible with LeetCode’s Python3 runtime; do NOT rely on undefined/implementation-specific behavior.
# - If you introduce any helper functions/classes, include their full implementations in this file; no hidden helpers.

class Solution(object):
    def kthSmallestProduct(self, nums1, nums2, k):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :type k: int
        :rtype: int
        """
        def count_less_equal(x):
            count = 0
            j = len(nums2) - 1
            for i in range(len(nums1)):
                while j >= 0 and nums1[i] * nums2[j] > x:
                    j -= 1
                count += (j + 1)
            return count

        left, right = -10**10, 10**10
        while left < right:
            mid = (left + right) // 2
            if count_less_equal(mid) < k:
                left = mid + 1
            else:
                right = mid
        return left

File: evaluation/S0/javascript/Hard/2120-kth-smallest-product-of-two-sorted-arrays.js
// Source: https://leetcode.com/problems/kth-smallest-product-of-two-sorted-arrays/   |   Difficulty: Hard
//
// Problem Description:
// Given two sorted 0-indexed integer arrays nums1 and nums2 as well as an integer k, return the kth (1-based) smallest product of nums1[i] * nums2[j] where 0 <= i < nums1.length and 0 <= j < nums2.length.
//
// Example:
// Input: nums1 = [2,5], nums2 = [3,4], k = 2
// Output: 8
// Explanation: The 2 smallest products are:
// - nums1[0] * nums2[0] = 2 * 3 = 6
// - nums1[0] * nums2[1] = 2 * 4 = 8
// The 2nd smallest product is 8.
//
// Constraints:
// 1 <= nums1.length, nums2.length <= 5 * 104
// 	-105 <= nums1[i], nums2[j] <= 105
// 	1 <= k <= nums1.length * nums2.length
// 	nums1 and nums2 are sorted.
//

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @param {number} k
 * @return {number}
 */
var kthSmallestProduct = function(nums1, nums2, k) {
    const countLessEqual = (x) => {
        let count = 0;
        let j = nums2.length - 1;
        for (let i = 0; i < nums1.length; i++) {
            while (j >= 0 && nums1[i] * nums2[j] > x) {
                j -= 1;
            }
            count += (j + 1);
        }
        return count;
    }
    let left = -1000000000;
    let right = 1000000000;
    while (left < right) {
        const mid = Math.floor((left + right) / 2);
        if (countLessEqual(mid) < k) {
            left = mid + 1;
        }
        else {            right = mid;
        }
    }
    return left;
};
